### PR TITLE
*: implement request header support

### PR DESCRIPTION
**Pull request by BusyJay**
This pr adds support to customize request headers. We need it to examine cluster id in TiKV.

close #135

### CODE REVIEW

File: grpc-sys/grpc_wrap.c [hunk 1]
```diff
@@ -160,55 +160,69 @@ grpcwrap_metadata_array_destroy_full(grpc_metadata_array *array) {
...
-grpcwrap_metadata_array_create(size_t capacity) {
-  grpc_metadata_array *array =
-      (grpc_metadata_array *)gpr_malloc(sizeof(grpc_metadata_array));
-  grpc_metadata_array_init(array);
-  array->capacity = capacity;
-  array->count = 0;
-  if (capacity > 0) {
-    array->metadata =
+GPR_EXPORT void GPR_CALLTYPE
+grpcwrap_metadata_array_init(grpc_metadata_array *array, size_t capacity) {
```
Comment (huachaohuang) [inline grpc-sys/grpc_wrap.c]: Should init `count` and `capacity` properly too.

**Comment by BusyJay** [comment]
PTAL

File: grpc-sys/grpc_wrap.c [hunk 2]
```diff
@@ -160,55 +160,69 @@ grpcwrap_metadata_array_destroy_full(grpc_metadata_array *array) {
...
   return (char *)GRPC_SLICE_START_PTR(array->metadata[index].value);
 }
 
+GPR_EXPORT void GPR_CALLTYPE
+grpcwrap_metadata_array_cleanup(grpc_metadata_array *array) {
+  grpcwrap_metadata_array_destroy_metadata_including_entries(array);
+}
+
+GPR_EXPORT void GPR_CALLTYPE
+grpcwrap_metadata_array_freeze(grpc_metadata_array *array) {
```
Comment (huachaohuang) [inline grpc-sys/grpc_wrap.c]: `freeze` seems not intuitive, consider using `shrink_to_fit` from C++ STL.

File: src/metadata.rs [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+
+use grpc_sys::{self, GrpcMetadataArray};
+use std::{ptr, slice, str};
+use std::borrow::Cow;
+
+use libc;
+
+use error::{Error, Result};
+
+fn normalize_key(key: &str, binary: bool) -> Result<Cow<str>> {
```
Comment (huachaohuang) [inline src/metadata.rs]: Is this required by gRPC? Any reference?

File: src/metadata.rs [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+///
+/// - Response trailers
+///
+///     They are sent by the server at the end of a remote call along with
+///     resulting call status.
+///
+/// Metadata value can be ascii string or bytes. They are distinguish by the
+/// key suffix, key of bytes value should have suffix '-bin'.
+#[repr(C)]
+pub struct Metadata(GrpcMetadataArray);
```
Comment (huachaohuang) [inline src/metadata.rs]: I think it is better to name it `MetadataArray` to keep the naming consistency with gRPC, it is weird that `Metadata` actually contains an array of metadata.

File: src/metadata.rs (hunk 3 above)
Comment (BusyJay) [reply src/metadata.rs]: See docs at https://github.com/pingcap/grpc-rs/pull/161/files#diff-26aa3c8895057aa93d60e92c0dee6ddcR145

File: src/metadata.rs (hunk 4 above)
Comment (BusyJay) [reply src/metadata.rs]: Actually its name is taken from [gRPC cshar](https://github.com/grpc/grpc/blob/master/src/csharp/Grpc.Core/Metadata.cs#L37).

File: tests/cases/metadata.rs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        .unwrap()
+        .add_bytes("k1-bin", &[0x00, 0x01, 0x02])
+        .unwrap();
+    let metadata = builder.build();
+    let call_opt = CallOption::default().headers(metadata);
+
+    let mut req = HelloRequest::new();
+    req.set_name("world".to_owned());
+    let resp = client
+        .say_hello_async_opt(&req, call_opt)
```
Comment (huachaohuang) [inline tests/cases/metadata.rs]: Why not just `say_hello_opt`?

**Comment by BusyJay** [comment]
@siddontang @overvenus @huachaohuang PTAL

**Comment by huachaohuang** [comment]
LGTM

**Comment by BusyJay** [comment]
PTAL

**Pull request merged**